File: src/Train.hs
module Main
       where

import Common      (getEyesMouth, readAnswers, readImages, rotImgCorrect,
                    shuffle, train, writePGM)
import System.Exit (exitSuccess)

main :: IO ()
main = do
    inputContents <- readFile "training.txt"
    inputAnswers <- readFile "training-facit.txt"
    let images = map (rotImgCorrect 20) $ readImages inputContents
        answers = readAnswers inputAnswers
        (eyeAnswers, mouthAnswers) = unzip (map getEyesMouth answers)
        eyes = zip images eyeAnswers
        mouth = zip images mouthAnswers

    shuffled <- shuffle (zip eyes mouth)

    let (shuffledEyes,shuffledMouth) = unzip shuffled
        (testEyes,trainEyes) = splitAt 50 shuffledEyes
        (testMouth,trainMouth) = splitAt 50 shuffledMouth


    eyeweights2 <- train trainEyes testEyes
    writePGM "eyew.pgm" eyeweights2
    writeFile "eye-weights" $ show eyeweights2
    mouthweights2 <- train trainMouth testMouth
    writePGM "mouthw.pgm" mouthweights2
    writeFile "mouth-weights" $ show mouthweights2

    exitSuccess

File: matlab/train.m
function [ew,mw ] = train(images,ew,mw,lr)


for j=1:length(images)

		ey = sigmoid(ew*images{j}.data);
		my  = sigmoid(mw*images{j}.data);
        
        delta = (images{j}.eye-ey);
        ew = ew + lr*delta.*images{j}.data';
        
        delta = (images{j}.mouth-my);
        mw = mw + lr*delta.*images{j}.data';

end

end

